### ISSUE TITLE
Wesnoth crash when units have negative hitpoints: overflow in ai attack calculations

### ISSUE THREAD

**Issue by ePaul** [wesnoth/wesnoth#3042]
I'm running v1.14.0+dev (43f30cdd613-Clean) (self-compiled), with the UMC "A troop in a war". Somehow I got enemy units with negative hitpoints (which is likely [a bug in that addon](https://forums.wesnoth.org/viewtopic.php?f=8&t=41451&p=627469#p627469)), but then Wesnoth simply crashes without any indication why.

**Reproduction**

* (possibly: install "A troop in a war" 0.1.6d, not sure if this is needed for reproducing.)
* Load this savegame [AG-Skirmish Runde 3-c.gz](https://github.com/wesnoth/wesnoth/files/1977548/AG-Skirmish.Runde.3-c.gz)
* Click "End turn".
→ Wesnoth crashes (window disappears, exit code 134).

The following is printed on the console (stdout/stderr):

```
Battle for Wesnoth v1.14.0+dev (43f30cdd613-Clean)
Started on Sun May  6 12:43:29 2018

Automatically found a possible data directory at /home/USER/projektoj/wesnoth

Data directory:               /home/USER/projektoj/wesnoth
User configuration directory: /home/USER/.config/wesnoth
User data directory:          /home/USER/.local/share/wesnoth/1.14
Cache directory:              /home/USER/.cache/wesnoth

Setting mode to 1920x1026
Checking lua scripts... ok
20180506 12:43:36 error display: Tile at 16,7 isn't on the map, can't scroll to the tile.
20180506 12:43:40 warning ai/general: overflow in ai attack calculation
20180506 12:43:40 warning ai/general: overflow in ai attack calculation
20180506 12:43:40 warning ai/general: overflow in ai attack calculation
20180506 12:43:40 warning ai/general: overflow in ai attack calculation
[... some thousand more of those lines ...]
20180506 12:43:40 warning ai/general: overflow in ai attack calculation
20180506 12:43:40 warning ai/general: overflow in ai attack calculation
20180506 12:43:40 warning ai/general: overflow in ai attack calculation
wesnoth: src/attack_prediction.cpp:1764: double {anonymous}::calculate_probability_of_debuff(double, bool, double, double, bool, double): Assertion `initial_prob >= 0.0 && initial_prob <= 1.0' failed.
Abgebrochen (Speicherabzug geschrieben)
```

(I didn't find the mentioned core dump, not sure where that would be.)

**Expected behavior**
If WML content creates an invalid state, Wesnoth should show a useful error message, and potentially go back to the main screen, not simply crash.

**Comment by ePaul** [comment]
When editing this save-game to not have negative hitpoints, the issue doesn't happen.

**Comment by ePaul** [comment]
My opinion (which doesn't necessarily count, as I'm not a developer) would be to give a useful error message when trying to load this. Though an alternative would be to just have a graceful fallback when the AI calculation in question here encounters this.

**Comment by gfgtdf** [comment]
afaik the engine itself also creates units with negative hitpoints during an attack, so disallowing setting umc negative hp woudl result in codes like `unit.hitpoints = unit.hitpoints` actually changing the hitpoints if it happens during an attack related event where a unit has negative hitpoints, which is very counterintuitive.

**Comment by CelticMinstrel** [comment]
Uhh, why would the engine be creating units with negative hitpoints?

**Comment by Vultraz** [comment]
I think what he means is units may have negative hitpoints after an attack when events like attack_end fire. Pretty sure I've observed this myself.

**Comment by CelticMinstrel** [comment]
That seems rather counter-intuitive… is there a reason for it?

**Comment by gfgtdf** [comment]
> I think what he means is units may have negative hitpoints after an attack when events like attack_end fire. Pretty sure I've observed this myself.

yes thats what i meant.

> That seems rather counter-intuitive… is there a reason for it?

note sure, but umc might rely on it to detemine how much damage has 'overflowed', i have seen umc doing crazy things.

**Comment by CelticMinstrel** [comment]
Yeah, and I could see providing that information to be a valid reason for this unintuitive behaviour. It should definitely be documented on the wiki though if it's not already.

**Comment by jyrkive** [comment]
All right, debugging shows why exactly negative HP ends up triggering this assertion.

[Omae wa mou shindeiru](http://knowyourmeme.com/memes/you-are-already-dead-omae-wa-mou-shindeiru).

In this savegame, the AI simulates a situation where unit A attacks unit B. Unit B has a poisoning attack, but its HP is negative. Meanwhile, unit A is so close to level-up that killing unit B will level up unit A and heal the poison status. Damage calculation takes both the poisoning attack and the upcoming level-up into account when calculating the probability that unit A will be poisoned after the attack.

The problem is that due to the negative HP, damage calculation code thinks that B is already dead before the fight (HP <= 0), but that B may not be dead after the fight (HP == 0). Thus, the probability that A kills B is considered to be **negative**, and the fact that killing B heals poison is considered to **increase** the probability that A is poisoned after the fight. Thus, the "A is poisoned" probability creeps to 105 %, which causes the crash if the AI simulates another fight for A after this one.

Next, time to start writing the fix...

**Comment by CelticMinstrel** [comment]
How could it break addons?

**Comment by gfgtdf** [comment]
the engine generates units with negative hp, the umc wml unstores it, then his change breaks to main menu (which is a really strange way to handle invalid umc api parameters in the first place)

**Comment by CelticMinstrel** [comment]
Ah. Yeah, I'd say at least it should just zero the hitpoints if they're found to be negative (or something similar).

**Comment by jyrkive** [comment]
@gfgtdf Addressed in commit dc6818799952c7b76e78c7309a30f8a6a6a7010c.

**Comment by jyrkive** [comment]
@gfgtdf Commit b0c1382693f4d72d7ad8f662ad91612da72b6781 addresses those four events. As far as I can tell, these six are the only events which can be called when the engine has set HP to a negative value.
